utils/disk_space_monitor.hh

```diff
@@ -0,0 +1,127 @@
+/*
+ * Copyright (C) 2024-present ScyllaDB
+ */
+
+/*
+ * SPDX-License-Identifier: AGPL-3.0-or-later
+ */
+
+#pragma once
+
+#include <chrono>
+#include <filesystem>
+#include <set>
+
+#include <boost/intrusive/list.hpp>
+
+#include <seastar/core/gate.hh>
+#include <seastar/core/lowres_clock.hh>
+#include <seastar/core/timer.hh>
+#include <seastar/util/optimized_optional.hh>
+
+#include "seastarx.hh"
+
+using namespace std::chrono_literals;
+
+namespace bi = boost::intrusive;
+
+namespace utils {
+
+// Instantiated only on shard 0
+class disk_space_monitor {
+public:
+    struct config {
+        scheduling_group sched_group;
+        lowres_clock::duration normal_polling_interval = 10s;
+        lowres_clock::duration high_polling_interval = 1s;
+        // Use high_polling_interval above this threshold
+        float polling_interval_threshold = 0.95;
+    };
```

Comment:
The file system can start behaving very very bad at 95%, failing allocations, etc. We're very sensitive to latency and latency can increase from the XFS side. Shouldn't we treat 90% as the upper bound? Otherwise, we're saying that we treat range (0.9, 0.95) as acceptable.

Reply:
If we're pushing towards 90% density, it means when we cross it, balancer will start migrating tablets away from the node. So the disk monitor, which is a replica entity, should treat 80-85% usage as high, so it can do local things to avoid hitting the balancer threshold.

Reply:
we can lower this interval, though migration should kick-in earlier than that (and it happens externally, using the node-exporter, not via this `disk_space_monitor`)

Reply:
yes, I was more worried about this monitor not having an actual effect, since balancer might kick in earlier. btw, we already call each node periodically to pull per-table stats for split / merge decisions, the interface can be extended too include per-node stats in additional to per-table ones I think.

Reply:
see storage_service::load_stats_for_tablet_based_tables(), this is the handler

Reply:
These should all be config items

Reply:
> we already call each node periodically to pull per-table stats for split / merge decisions, the interface can be extended too include per-node stats in additional to per-table ones I think.

Let's wait with that until we do something with this information.
I'm not sure that extending load_stats_for_tablet_based_tables makes sense since it provides per-table stats.
I think that a new rpc would be more appropriate.

Reply:
> > we already call each node periodically to pull per-table stats for split / merge decisions, the interface can be extended too include per-node stats in additional to per-table ones I think.
> 
> Let's wait with that until we do something with this information. I'm not sure that extending load_stats_for_tablet_based_tables makes sense since it provides per-table stats. I think that a new rpc would be more appropriate.

```
struct load_stats {
    std::unordered_map<table_id, table_load_stats> tables;

    load_stats& operator+=(const load_stats& s);
    friend load_stats operator+(load_stats a, const load_stats& b) {
        return a += b;
    }
};
```

we feed load_stats into the balancer (pulled periodically by the coordinator), so it makes sense to extend load_stats with map<host_id, node_load_stats>. but anyway, we don't have to make a decision now.

Reply:
we could reuse it in the future for cpu balancer, etc.